upup/pkg/fi/cloudup/dns.go

```diff
@@ -150,18 +150,22 @@ func precreateDNS(cluster *api.Cluster, cloud fi.Cloud) error {
 		return fmt.Errorf("error getting DNS resource records for %q", zone.Name())
 	}
 
-	// TODO: We should change the filter to be a suffix match instead
-	//records, err := rrs.List("", "")
-	records, err := rrs.List()
-	if err != nil {
-		return fmt.Errorf("error listing DNS resource records for %q: %v", zone.Name(), err)
-	}
-
 	recordsMap := make(map[string]dnsprovider.ResourceRecordSet)
-	for _, record := range records {
-		name := dns.EnsureDotSuffix(record.Name())
-		key := string(record.Type()) + "::" + name
-		recordsMap[key] = record
+	// vSphere provider uses CoreDNS, which doesn't have rrs.List() function supported.
+	// Thus we use rrs.Get() to check every dnsHostname instead
+	if cloud.ProviderID() != "vsphere" {
```

Comment:
Instead of hardcoding "vsphere", better to use the const CloudProviderVSphere defined in cloud.go.